pages/outputs/[id].jsx

```diff
@@ -0,0 +1,149 @@
+
+const LOCALE = 'en-GB'
+const CITATION_STYLES = ['apa', 'bibtex']
+
+// Needed for clear development of the data retrieval
+// At the time of development the API was unstable
+const log = (...args) => {
+  if (process.env.NODE_ENV !== 'production')
+    // eslint-disable-next-line no-console
+    console.log(...args)
+}
+
+export async function getStaticProps({ params: routeParams }) {
+  const { id } = routeParams
+
+  const data = {}
+  let revalidate = 60 * 60 * 24 * 7 // seconds, i.e. every week
+
+  try {
+    const rawOutput = await fetchMetadata(id)
+
+    // Strip some properties to optimise network traffic
+    const { fullText: _, ...output } = rawOutput
+
+    const { data: dataProvider } = await request(output.dataProvider)
+    Object.assign(data, {
+      ...output,
+      dataProvider,
+    })
+  } catch (error) {
+    log(error)
+
+    return {
+      props: { error },
+      notFound: true,
+    }
+  }
+
+  try {
+    const doi = data.identifiers?.doi
+    if (!doi) throw new Error('No DOI — no citation')
+
+    const citations = await fetchCitations(doi, {
+      styles: CITATION_STYLES,
+      locale: LOCALE,
+    })
+    data.citations = citations
+  } catch (citationRetrievalError) {
+    log(citationRetrievalError)
+
+    // If any error happens, we pretend, citation could not be generated
+    // or retrieved. This behaviour should be improved because we could generate
+    // citations ourselves using some library.
+    //
+    // Take your time to explore:
+    // - https://citeproc-js.readthedocs.io/en/latest/index.html
+    // - https://citationstyles.org
+    data.citations = []
+  }
+
+  try {
+    const similarOutputs = await fetchSimilarTo(id)
+
+    // Strip some properties to optimise network traffic
+    data.similarOutputs = similarOutputs.map(
+      ({ fullText, ...output }) => output
+    )
+
+    // TODO: Remove this after the API fix
+    data.similarOutputs = data.similarOutputs.map((output) => ({
+      ...output,
+      authors: output.authors.map(({ name }) => name),
+    }))
+
+    if (typeof data.similarOutputs[0].links[0] != 'object') {
+      // eslint-disable-next-line no-console
+      console.log(data.similarOutputs[0].links)
+      throw new Error('Links are not properly formatted at the moment...')
+    }
+  } catch (error) {
+    log(error)
+
+    // If any error happens, we pretend, there were no recommendations
+    // This behaviour could be changed to explicit error reporting but should
+    // be considered deeper.
+    data.similarOutputs = []
+    revalidate = 30
+  }
+
+  return {
+    props: { data },
+    revalidate,
+  }
+}
+
+export async function getStaticPaths() {
+  // Generating only 1st page as static for the first time
+  // In the future we could retrieve a number of most popular pages
+  // and pre-render them to static
+  const paths = [
+    {
+      params: { id: '1' },
+    },
+  ]
+
+  return { paths, fallback: 'blocking' }
+}
```

Comment:
However, if you use `getStaticProps`, you need to have `getStaticPath` — a function that returns the list of paths to generate static pages from.

The returning object must include `paths` property, that is an array of `params`, which contains only `id` in our case, as I said before.

The `fallback` property, in this case, means, that the page must be rendered before returning to the user if it wasn't rendered before. Other options are "return error always" or "return error first time but render later", which don't fit us.

So, effectively I pre-render the page with `id=1` and render every other page on the server side when it's requested. Next.js handle cache for me.